### PR TITLE
Show name for open generic arguments

### PR DESCRIPTION
**Pull request by AndreyTretyak**
- Change open generics display from `List<>` to `List<T>`.
- Add unit test for open generics.
- Add a short name for `void` type and test for it.
- Fix variable name typo.

@smitpatel could you please check if it's in line with what you offered here #278.

### CODE REVIEW

**Comment by smitpatel** [comment]
@Eilon - Your call on what should open generic print.

**Comment by AndreyTretyak** [comment]
A small point against option with the name. If you will have a class with name `T` (surely not the best naming) there is no way to say if `List<T>` open or close generic.

**Comment by smitpatel** [comment]
`c.IsGenericParameter` is true for generic parameter false for non-generic.

**Comment by AndreyTretyak** [comment]
I meant, there is no way to distinguish in string representation that we get from TypeNameHelper

**Comment by Eilon** [comment]
My preference would be to avoid the `<TKey, TValue>` and instead do `<,>`. I think it's clear enough what the type is.

File: shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs [hunk 1]
```diff
@@ -51,13 +52,14 @@ private static void ProcessType(StringBuilder builder, Type type, bool fullName)
...
-            else if (_builtInTypeNames.TryGetValue(type, out var buildInName))
+            else if (_builtInTypeNames.TryGetValue(type, out var builtInName))
             {
-                builder.Append(buildInName);
+                builder.Append(builtInName);
             }
             else
             {
-                builder.Append(fullName ? type.FullName : type.Name);
+                var name = fullName || type.IsGenericParameter ? type.FullName : type.Name;
```
Comment (smitpatel) [inline shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs]: If `type.IsGenericParameter` then skip fully.
for typeof(List<>), we want to print `List<>` for full name and short name both. With this code shortname will print `List<T>`

File: shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs [hunk 2]
```diff
@@ -115,9 +117,15 @@ private static void ProcessGenericType(StringBuilder builder, Type type, Type[]
...
             {
                 ProcessType(builder, genericArguments[i], fullName);
-                if (i + 1 != length)
+                if (i + 1 == length)
+                {
+                    continue;
+                }
+
+                builder.Append(',');
+                if (!genericArguments[i + 1].IsGenericParameter)
```
Comment (smitpatel) [inline shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs]: If one parameter is generic then all of them are generic. Check for one and just do string join for rest and skip looping fully.

File: test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs [hunk 3]
```diff
@@ -9,110 +9,124 @@ namespace Microsoft.Extensions.Internal
...
+            {
+                "Microsoft.Extensions.Internal.TypeNameHelperTest+Level1<int>+Level2<bool>+Level3<int>",
+                typeof(Level1<int>.Level2<bool>.Level3<int>)
+            };
         }
 
-        [Fact]
-        public void Can_pretty_print_CLR_name()
+        [Theory]
+        [MemberData(nameof(GetFullTypeNamesTestData))]
```
Comment (smitpatel) [inline test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: Either put `InlineData` or just keep the tests as is. Not a fan of `MemberData` for readability purposes.

File: test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs [hunk 4]
```diff
@@ -128,6 +142,66 @@ public void Can_pretty_print_array_name(string expected, Type type, bool fullNam
...
+
+            yield return new object[]
+            {
+                "System.Collections.Generic.Dictionary<Microsoft.Extensions.Internal.TypeNameHelperTest+B<>,>",
+                closedDictionaryType
+            };
+
+            var openLevelType = typeof(Level1<>.Level2<>.Level3<>);
+            var genArgsLevel = openLevelType.GetGenericArguments();
+            genArgsLevel[1] = typeof(string);
```
Comment (smitpatel) [inline test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: If this types are not constructible without reflection then leave it. We don't need tests for that. Test may verify the pipeline but no point of writing tests which are impossible without ugly hacks.

File: shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs (hunk 2 above)
Comment (AndreyTretyak) [reply shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs]: There are cases when one parameter open and another close, for example:
```
class TestType<T> : BaseType<int,T>
{
}
```

File: test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs (hunk 3 above)
Comment (pakrym) [reply test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: If arguments are too big for `InlineData` we use `MemberData` in many places, it's much better then putting all tests cases in one `[Fact]`

File: shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs (hunk 1 above)
Comment (AndreyTretyak) [reply shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs]: For short name this code also will print `List<>`, there is a unit test for it [here](https://github.com/aspnet/Common/pull/283/files#diff-65ffadf8398a22a547ef1fc2cf411cb5R135), but I see your point, skipping fully would be easier.

File: shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs (hunk 2 above)
Comment (smitpatel) [reply shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs]: typeof(TestType<>) has only 1 generic parameter.

Comment (smitpatel) [reply shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs]: If it is possible to create, it is good logic, but we should also add test for it.

File: test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs (hunk 4 above)
Comment (AndreyTretyak) [reply test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: Those types totally possible without reflection, using inheritance. It's just easier to construct them it that way than creating class hierarchy for each one.

File: test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs (hunk 3 above)
Comment (AndreyTretyak) [reply test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: But If we make it a single test then execution will stop after first error and we won't get info about all of them, also inside `GetFullTypeNamesTestData` every test case clearly distinguishable and readable and it`s near with test itself. I'll try to put it in the `InlineData`, but some lines could be quite bulky.

Comment (smitpatel) [reply test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: Leave MemberData as is :)

File: shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs (hunk 2 above)
Comment (AndreyTretyak) [reply shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs]: `typeof(TestType<T>).BaseType` will have one open and one close parameter (`int` and `T`), unit test [here](https://github.com/aspnet/Common/pull/283/files#diff-65ffadf8398a22a547ef1fc2cf411cb5R152) done to test this logic.

File: test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs [hunk 5]
```diff
@@ -9,110 +9,124 @@ namespace Microsoft.Extensions.Internal
...
-                TypeNameHelper.GetTypeDisplayName(typeof(int)));
-            Assert.Equal("System.Collections.Generic.List<int>",
-                TypeNameHelper.GetTypeDisplayName(typeof(List<int>)));
-            Assert.Equal("System.Collections.Generic.Dictionary<int, string>",
-                TypeNameHelper.GetTypeDisplayName(typeof(Dictionary<int, string>)));
-            Assert.Equal("System.Collections.Generic.Dictionary<int, System.Collections.Generic.List<string>>",
-                TypeNameHelper.GetTypeDisplayName(typeof(Dictionary<int, List<string>>)));
-            Assert.Equal("System.Collections.Generic.List<System.Collections.Generic.List<string>>",
-                TypeNameHelper.GetTypeDisplayName(typeof(List<List<string>>)));
+            yield return new object[] { "int", typeof(int) };
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: More of a nit, you could return `TheoryData` instead. It removes some of the cruft around declaring these:

```C#
public static TheoryData GetFullTypeNamesTestData()
{ 
   return new TheoryData<Type, string>
   {
      { typeof(int), "int" },
      ...
    };
}
```

Also generally, inputs appear first in the test method arg list followed by expectations.

File: test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs (hunk 3 above)
Comment (AndreyTretyak) [reply test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: Thanks, I've tried option with `InlineData` and it actually not that bad, if [this](https://gist.github.com/AndreyTretyak/57623e99c4596c6f16ae8109edaceee7) is acceptable I can commit it.

File: test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs [hunk 6]
```diff
@@ -128,6 +142,66 @@ public void Can_pretty_print_array_name(string expected, Type type, bool fullNam
...
+            };
+
+            var openLevelType = typeof(Level1<>.Level2<>.Level3<>);
+            var genArgsLevel = openLevelType.GetGenericArguments();
+            genArgsLevel[1] = typeof(string);
+            var closedLevelType = openLevelType.MakeGenericType(genArgsLevel);
+
+            yield return new object[]
+            {
+                "Microsoft.Extensions.Internal.TypeNameHelperTest+Level1<>+Level2<string>+Level3<>",
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: How does one construct such a type without reflection? I thought C# doesn't let you mix open and closed generic types this way.

File: test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs (hunk 5 above)
Comment (AndreyTretyak) [reply test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: Thank you, I did know about that, look much better with concrete types and without arrays of objects.

File: test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs (hunk 6 above)
Comment (AndreyTretyak) [reply test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: For case like this `class TestType<T> : BaseType<int,T> { ... }` result of `typeof(TestType).BaseType` will be type with one open and one closed parameter. C# doesn't allow mix generics only in `typeof` operator.

Comment (AndreyTretyak) [reply test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs]: I'll add one unit test for such type without reflection to make it more apparent.

**Comment by pranavkm** [comment]
:shipit:

**Comment by pranavkm** [comment]
@smitpatel is this good to go? This change is blocking another PR I'm looking at and I'd like to get this in to unblock it.

**Comment by smitpatel** [comment]
@pranavkm - Go ahead.

**Comment by pranavkm** [comment]
Thanks for the PR!

**Comment by pranavkm** [comment]
@Eilon so an unfortunate side effect of our choice to omit the generic parameter name is that our stack traces look bad. We get 

`Microsoft.Extensions.Internal.StackTraceHelperTest+GenericClass<>.GenericMethod<>(ref  value)`

instead of 

`Microsoft.Extensions.Internal.StackTraceHelperTest+GenericClass<T>.GenericMethod<V>(ref V value)",`

Ignoring the extra white space produced in the first stack trace, the second one just conveys the type info a lot better. Should we reconsider the choice to not print generic parameter name?

**Comment by AndreyTretyak** [comment]
Maybe it could be done as an option? An output of generic parameters names could be very useful for method display in #270 or you already have a way for method display and those PR no longer relevant?

**Comment by pranavkm** [comment]
@AndreyTretyak I'd prefer not having a flag for it. Given that this is a Sources package, it's just going to push the decision further up the stack rather than choising useful defaults

**Comment by AndreyTretyak** [comment]
I understand. Could we hide open generics for type display but show them in methods where it's really needed, or it will be too inconsistent? 
If we add a method into `TypeNameHelper` for methods display,  we could hide all logic about selecting correct variant inside it. 
Then we could use those method in `StackTraceHelper.GetMethodDisplayString` in DI exception with a list of constructors and other places.

**Comment by pranavkm** [comment]
I'm trying to determine what @Eilon's rationale was in skipping it originally.

**Comment by smitpatel** [comment]
@pranavkm - `typeof(List<>)`  prints out `List<>` . The output is as you would write in source code.

**Comment by Eilon** [comment]
Yeah what @smitpatel said... not sure the best way to address this.

**Comment by Eilon** [comment]
I'm fine adding a new optional parameter (with the default being the "old" behavior).

**Pull request merged**